agent/command/archive_util.go

```diff
@@ -222,14 +221,137 @@ func TarGzReader(path string) (f, gz io.ReadCloser, tarReader *tar.Reader, err e
 	return f, gz, tarReader, nil
 }
 
-// TarGzWriter returns a file, gzip writer, and tarWriter for the path.
+// tarGzWriter returns a file, gzip writer, and tarWriter for the path.
 // The tar writer wraps the gzip writer, which wraps the file.
-func TarGzWriter(path string) (f, gz io.WriteCloser, tarWriter *tar.Writer, err error) {
+func tarGzWriter(path string, useParallelGzip bool) (f, gz io.WriteCloser, tarWriter *tar.Writer, err error) {
 	f, err = os.Create(path)
 	if err != nil {
 		return nil, nil, nil, errors.Wrapf(err, "creating file '%s'", path)
 	}
-	gz = gzip.NewWriter(f)
+	if useParallelGzip {
+		gz = pgzip.NewWriter(f)
+	} else {
+		gz = gzip.NewWriter(f)
+	}
 	tarWriter = tar.NewWriter(gz)
 	return f, gz, tarWriter, nil
 }
+
+// archiveContentFile represents a tar file on disk.
+type archiveContentFile struct {
+	path string
+	info os.FileInfo
+	err  error
+}
+
+// findContentsToArchive finds all files starting from the rootPath with the
+// given inclusion and exclusion patterns.
+func findContentsToArchive(ctx context.Context, rootPath string, includes, excludes []string) (files []archiveContentFile, totalSize int, err error) {
+	out := []archiveContentFile{}
+	catcher := grip.NewBasicCatcher()
+	archiveContents, totalSize, err := streamArchiveContents(ctx, rootPath, includes, excludes)
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "getting archive contents")
+	}
+	for _, fn := range archiveContents {
+		if fn.err != nil {
+			catcher.Add(fn.err)
+			continue
+		}
+
+		out = append(out, fn)
+	}
+
+	if catcher.HasErrors() {
+		return nil, 0, catcher.Resolve()
+	}
+
+	return out, totalSize, nil
+}
+
+func streamArchiveContents(ctx context.Context, rootPath string, includes, excludes []string) (files []archiveContentFile, totalSize int, err error) {
```

Comment:
It's a little hard to tell what changed since you moved files around -- I'd generally suggest avoiding big refactors when making changes to simplify reviewing. It seems here like we only added addUniqueFile in order to compute totalSize, to use later in making a distnction for pgzip? Could we add a comment here to that effect?

Reply:
Eh actually disregard, the refactor is obviously a good one. A comment for the inner function would still be helpful I think though.